src/realm/sort_descriptor.cpp

```diff
@@ -434,20 +476,33 @@ bool BaseDescriptor::Sorter::operator()(IndexPair i, IndexPair j, bool total_ord
                 const auto& obj = m_columns[t].table->get_object(key_i);
                 const auto& col_key = m_columns[t].col_key;
 
-                cache_i.value = col_key.get_value(obj);
+                // store stringID instead of the actual string if possible
+                cache_i.cached_string_id = col_key.get_string_id(obj);
```

Comment:
This will add unnecessary overhead per comparison when sorting on a non-string column. Ideally instead of doing the check inside the loop, we would have two separate specialized loops to keep other types fast. This might be complex to refactor though because there are already multiple optimizations piled in here. 
ie. instead of this pattern:
```
for (..) {
if interned string 
    do special comparison
else 
    do value comparison
}
```
we should do something like this:
```
if column supports string interning
    for (...) {
        do special intern comparison if possible
    }
} else {
    for (...) {
        do normal value comparison
    }
}
```
I think we should have some benchmarks on sorting integers that can back up my claim, I wonder how much this change costs them?

Reply:
The cost of fetching the ID for a non-mixed or string will be only check the column ID + an extra check for the optional type, so I don't expect this to be huge. However, you have a point, we can handle this only for strings or string like properties.

Reply:
Yep, I think statically it does not make a huge difference, probably we are going to slow down a little bit Int sorting, but, it is really difficult to gauge and also doesn't it depend on also by the data we have?

```
Req runs:    6  SortInt (MemOnly, EncryptionOff):             min  57.16ms (+3.73%)            max  76.90ms (+30.61%)           med  66.13ms (+17.93%)           avg  66.87ms (+18.40%)           stddev  8.98ms (+586.99%)     

Req runs:    5  SortIntList (MemOnly, EncryptionOff):         min 321.11ms (+15.11%)           max 397.76ms (+27.72%)           med 333.63ms (+9.98%)            avg 353.77ms (+18.18%)           stddev 36.49ms (+194.74%)     

Req runs:   10  SortIntDictionary (MemOnly, EncryptionOff):     min  37.68ms (-0.91%)            max     56ms (+7.95%)            med  40.05ms (-5.75%)            avg  42.62ms (-1.18%)            stddev  6.44ms (+38.90%)      

Req runs:  225  SortThenLimit (MemOnly, EncryptionOff):         min   1.89ms (+14.04%)           max   3.82ms (+44.19%)           med   2.09ms (+21.53%)           avg   2.17ms (+23.60%)           stddev   284us (+144.66%) 
```

Reply:
This looks like a pretty big perf regression to me?

Reply:
There was a bug in my code, I was constantly passing by copy the Extended Column, these are the right perfs:

```
Req runs:    8  SortInt (MemOnly, EncryptionOff):             min  58.66ms (+6.49%)            max  66.55ms (+11.33%)           med  60.24ms (+4.87%)            avg  61.30ms (+6.42%)            stddev  3.06ms (+70.90%)      

Req runs:    5  SortIntList (MemOnly, EncryptionOff):         min 290.56ms (+3.83%)            max 304.51ms (-1.68%)            med 302.37ms (+3.53%)            avg 299.28ms (+1.51%)            stddev  6.12ms (-45.79%)      

Req runs:   10  SortIntDictionary (MemOnly, EncryptionOff):     min  37.89ms (-3.70%)            max  45.80ms (-12.86%)           med  41.27ms (-0.43%)            avg  41.59ms (-3.61%)            stddev  2.53ms (-36.80%)      

Req runs:  231  SortThenLimit (MemOnly, EncryptionOff):         min   1.84ms (+10.23%)           max   2.93ms (-24.94%)           med   1.95ms (+6.03%)            avg   1.98ms (+1.44%)            stddev   144us (-58.33%)      

```

I am still failing to see why this is a massive problem for non-mixed or string columns. It does not seem like it to me, but I must be wrong. Because you are both complaining about this :-) 

I agree that Mixed and Strings are going to be impacted though.

Reply:
What is the benchmark difference of `BenchmarkSort` (the name is "Sort") of this branch compared to master? AFAICT that is the benchmark that actually sorts strings.

Reply:
Yep, I tried `BenchmarkSort` already. And it is slower. But at the same time it is not the best benchmark either IMO. 

`
Req runs:    5  Sort (MemOnly, EncryptionOff):              * min 207.91ms (+72.16%)           max 214.60ms (+48.24%)         * med 210.27ms (+62.98%)         * avg 211.07ms (+60.69%)           stddev  2.99ms (-65.96%)      
`

Mainly for these 3 reasons:

1. The internal DS that the string interner is keeping is not lock free, we are constantly grabbing a lock. Especially for sorting stuff, it is a real problem (since we may be calling the sorting operator a lot) . Jira to cover this: (https://jira.mongodb.org/browse/RCORE-2214)

2. We are constantly loading a lot of data when we need to access the compressed string leaves in the Trie we built, we need to improve the internal DS: (https://jira.mongodb.org/browse/RCORE-2125)

3. `BenchmarkSort ` is writing 200k randomly generated strings with 1 character each. This is very much the worse scenario for interning stuff. Also, in the feature branch we are not limiting compression of strings like we do for integers (where we check if there is a sizeable gain before compressing). We just compress all. 

I've added a benchmark for testing the sorting of strings that are in compressed format (as you suggested). Passing 2 parameters to the test: 
- N the number of strings to store
- M the number of duplicated strings

 You can see that already, with all these limitations, there is a point after that it pays off to sort using the compressed string ID.

```
Req runs:    5  SortCompressedSmall(10,500) (MemOnly, EncryptionOff):   * min 120.99ms (+26.67%)           max 127.01ms (+4.03%)          * med 121.55ms (+11.37%)         * avg 123.02ms (+12.46%)           stddev  2.61ms (-75.99%)      

Req runs:    5  SortCompressedMedium(50,500) (MemOnly, EncryptionOff):     min 122.27ms (+8.63%)            max 146.98ms (+21.70%)           med 130.05ms (+14.46%)           avg 131.95ms (+14.63%)           stddev  9.44ms (+183.93%)     

Req runs:    5  SortCompressedLarge(100,500) (MemOnly, EncryptionOff):   * min 129.96ms (+17.27%)           max 135.25ms (+17.97%)         * med 133.83ms (+20.02%)         * avg 133.50ms (+18.78%)           stddev  2.14ms (+28.62%)      

This is much better, Each String is 1K chars and we have 5K of duplicated strings.
Req runs:    5  SortCompressedLarge(1000,5000) (MemOnly, EncryptionOff):     min 208.58ms (-26.71%)           max 222.46ms (-32.17%)           med 209.48ms (-27.31%)           avg 213.31ms (-27.96%)           stddev  6.25ms (-65.78%)      

```

Reply:
For the non-string benchmarks, we appear to be less than 10% worse which I think seems reasonable. Ideally, that code would not be affected at all, and there may be a way to do this by templating out the non-string sorting, but I'm not sure it is worth the duplication that will create.

For the BenchmarkSort case, I don't think we should ship this feature until we get that regression down by quite a bit. Can you create a ticket for the third item you suggested so that we don't forget about it? As long as we are tracking those 3 approaches and get them and recheck the benchmarks later, it shouldn't block this PR now.

Reply:
Yep I agree.
1. There is a bit of tech debt in the sorting logic. I tried to make the code templated based on column type, but I quit, because the code was basically duplicated, also we need to extract the column type in at least 2 cases. First one when we set the index pair and the other one when we are sorting the index pairs (too much hassle).

2. Yes I agree with you, in fact all this stuff is not going to next major. I will indeed add a new Jira for tracking the perf work (along the other tasks already mentioned).

Reply:
>  in fact all the code is not going into next major

What do you mean by this?

Reply:
I meant: that the code is not finished, we need to work on the optimizations I mentioned. The feature branch is just our base. Once we complete all the features, it will of course go into next major, like we did for integers.